crates/rolldown_binding/src/options/plugin/types/binding_js_or_regex.rs

```diff
@@ -1,29 +1,85 @@
+use std::fmt::Debug;
+
+use napi::{
+  bindgen_prelude::{FromNapiValue, Function, TypeName, ValidateNapiValue},
+  sys, Either, Env, Error, JsObject, JsUnknown, NapiValue, Status,
+};
+use serde::{Deserialize, Deserializer};
+
 use rolldown_utils::js_regex::HybridRegex;
 use rolldown_utils::pattern_filter::StringOrRegex;
-use serde::Deserialize;
 
-#[napi_derive::napi(object)]
 #[derive(Debug, Deserialize, Default, Clone)]
-#[serde(rename_all = "camelCase")]
-/// For String, value is the string content, flag is the `None`
-/// For Regex, value is the regular expression, flag is the `Some()`.
-/// Make sure put a `Some("")` in flag even there is no flag in regexp.
-pub struct BindingStringOrRegex {
-  pub value: String,
-  /// There is a more compact way to represent this, `Option<u8>` with bitflags, but it will be hard
-  /// to use(in js side), since construct a `JsRegex` is not used frequently. Optimize it when it is needed.
-  pub flag: Option<String>,
+pub struct JsRegExp {
+  pub source: String,
+  pub flags: String,
+}
+
+impl ValidateNapiValue for JsRegExp {}
+
+impl TypeName for JsRegExp {
+  fn type_name() -> &'static str {
+    "RegExp"
+  }
+
+  fn value_type() -> napi::ValueType {
+    napi::ValueType::Object
+  }
+}
+
+impl FromNapiValue for JsRegExp {
+  unsafe fn from_napi_value(env: sys::napi_env, napi_val: sys::napi_value) -> napi::Result<Self> {
+    let js_object = unsafe { JsObject::from_raw_unchecked(env, napi_val) };
+
+    let env = Env::from(env);
+    let global = env.get_global()?;
+    let regexp_constructor = global.get_named_property::<Function<JsUnknown, ()>>("RegExp")?;
+
+    if js_object.instanceof(regexp_constructor)? {
+      let source = js_object.get_named_property::<String>("source")?;
+      let flags = js_object.get_named_property::<String>("flags")?;
+
+      Ok(JsRegExp { source, flags })
+    } else {
+      Err(Error::new(Status::ObjectExpected, "Expect a RegExp object"))
+    }
+  }
+}
+
+#[derive(Debug, Clone)]
+pub struct BindingStringOrRegex(pub Either<String, JsRegExp>);
+
+impl Default for BindingStringOrRegex {
+  fn default() -> Self {
+    Self(Either::A(String::default()))
+  }
+}
```

Comment:
I don't see any `Default` usage

Reply:
This is because its parent object has `Default` trait, so we must implement the trait for it.

Reply:
Let me remove it because it doesn't seem necessary.